11x-new-kafka-roller.md

```diff
@@ -0,0 +1,263 @@
+
+We expect to remain in Alpha for 2 releases, then in Beta for at least 2 releases, if not more until we are happy that issues are ironed out and the roller is running stable.
+
+Reddit is one of the Strimzi vendors that offered to test the new roller while it's in Alpha and Beta phase.
+
+### Future improvement
+
+In the future, we can optimize Cruise Control's `BrokerSetAwareGoal` to make the batch rolling more efficient when the user enables Cruise Control in their clusters.
+This goal operates at the level of broker sets, which may correspond to physical or logical boundaries like racks, data centres, availability zones or custom logical groupings defined by administrators.
+It ensures that replicas of a partition are not assigned to the same broker set by spreading them across sets as evenly as possible.
+The new roller could then restart brokers in the same set in parallel while maintaining availability of partitions.
+This goal can be used for both rack aware and non rack aware clusters, compared to `RackAwareGoal` which relies on Kafka's built-in `broker.rack` property.
+
+
+This solution still has the limitation mentioned in the `Rejected Alternatives` section, that we can't be certain that other tooling hasn't reassigned some replicas since the last rebalance.
+In this case, the proposed algorithm can be used to check that brokers in the same set have no common partitions.
+This can be discussed further in the future.
+
+## Affected/not affected projects
+
+This proposal affects only
+the [`strimzi-kafka-operator`](https://github.com/strimzi/strimzi-kafka-operator).
+
+## Compatibility
+
+The new KafkaRoller introduced by this proposal will be used only for KRaft based clusters.
+The purpose of this proposal is to maintain the same behaviour of the old roller with a finite state machine design while adding the new features.
+
+## Rejected
+
+- Why not use rack information when batching brokers that can be restarted at the same time?
+When all replicas of all partitions have been assigned in a rack-aware way then brokers in the same rack trivially share no partitions, and so racks provide a safe partitioning.
+However nothing in a broker, controller or cruise control is able to enforce the rack-aware property therefore assuming this property is unsafe.
+Even if CC is being used and rack aware replicas is a hard goal we can't be certain that other tooling hasn't reassigned some replicas since the last rebalance, or that no topics have been created in a rack-unaware way.
+
+- Why not refactor the existing KafkaRoller rather than rewriting as a new component?
+It is also not particularly easy to unit test which resulted in insufficient test coverage for many edge cases, making it challenging to refactor safely.  The past experience of adding non trivial changes to it has proven to be expensive and more risky of introducing bugs because of tightly coupled logics and lack of testability.
```

Comment:
```suggestion
The existing roller is not particularly easy to unit test which has resulted in insufficient test coverage for many edge cases, making it challenging to refactor safely.
The past experience of adding non trivial changes to it has proven to be expensive, with a high risk of introducing bugs because of tightly coupled logics and lack of testability.
```